### ISSUE TITLE
Writing Tests

### ISSUE THREAD

**Issue by rhugga** [apenella/go-ansible#105]
Anyone know of a good example of writing tests for code using this library. Where I work we have a ridiculous pull request process that requires near total test coverage and I'm having a helluva time coming up with some meaningful tests. Especially since I'm wrapping the code inside the urfave/cli framework which in itself is a challenge to write tests for.

For example, here is a simple ping command that a user can run to validate their credentials and ansible and everything is working:
```
// actionPing is the wrapper function for the ping subcommand
func actionPing(mctx *context.Context, opts *common.AnsibleOptions) func(app *cli.Context) error {
	return func(app *cli.Context) error {

		// add timeout to our context
		ctx, cancel := context.WithTimeout(*mctx, time.Second*time.Duration(opts.Timeout))
		defer cancel()

		return ping(&ctx, opts)
	}
}

func ping(ctx *context.Context, opts *common.AnsibleOptions) error {

	// create buffer for our playbook callback
	buff := new(bytes.Buffer)

	// this populates all the required ansible fields from cli options
	pb := common.NewPlaybookCommand(opts, buff)

	// add playbooks to run
	pb.Playbooks = opts.Playbooks

	// check if we need sudo password, if so fetch it from tty
	if opts.Become {
		pb.PrivilegeEscalationOptions.Become = true
		password, err := utils.GetPassword(utils.GetPasswordFromStdin{Prompt: "Enter sudo password: "})
		if err != nil {
			return err
		}
		if err := pb.Options.AddExtraVar("ansible_sudo_pass", password); err != nil {
			return err
		}
	}

	// add our inventory
	pb.Options.Inventory = opts.Inventory

	// run the command
	err := pb.Run(*ctx)
	if err != nil {
		return err
	}

	return nil
}
```

Is there any way to mock the pb.Run() call? I don't really need to test/mock ansible, just that my function is working end to end.

Thanks for any help.

**Comment by apenella** [comment]
Hi @rhugga 
I use the following approach to test an application that uses go-ansible library. The main idea is testing how the library is called, and to achieve it, go-library can't called it directly.

To do that I wrapped the library in a facade which has a set of methods to configure the `AnsiblePlaybookCmd` object. I also created a mock for that wrapper, then I could inject either the wrapper or the mock to the object that executes the playbook. I use `github.com/stretchr/testify/mock` package to create mocks.

In the [stevedore](https://gostevedore.github.io/) project, where I use go-ansible library, I am working on a huge refactor where I introduced the described testing strategy. Take a look at https://github.com/gostevedore/stevedore/tree/version_0.11.0/internal/infrastructure/driver/ansible to an example.

I hope that it could help you!
